File: src/main/dart/net/coderextreme/data/TextExamples.dart
// Copyright 2021 by John Carlson.  All rights reserved.
import 'dart:io';
import '../x3d.dart';
var X3D0 =  X3D(

      profile_ : SFString('Immersive'),
      version_ : SFString('4.0'),
      head_ : 
        head(
          meta_ : [
            meta(
              name_ : SFString('title'),
              content_ : SFString('TextExamples.x3d')),

            meta(
              name_ : SFString('description'),
              content_ : SFString('Show different escape-character text examples for embedded quotation marks.')),

            meta(
              name_ : SFString('creator'),
              content_ : SFString('Don Brutzman')),

            meta(
              name_ : SFString('created'),
              content_ : SFString('7 April 2001')),

            meta(
              name_ : SFString('modified'),
              content_ : SFString('26 April 2016')),

            meta(
              name_ : SFString('warning'),
              content_ : SFString('Note that X3D Canonicalization (C14N) will scrub alternate XML character representations, be careful to check original encoding into version control.')),

            meta(
              name_ : SFString('warning'),
              content_ : SFString('Usually this source document needs to be inspected and edited using a plain-text editor in order to see the differences in these XML-equivalent text representations.')),

            meta(
              name_ : SFString('identifier'),
              content_ : SFString('https://www.web3d.org/x3d/content/examples/Basic/development/TextExamples.x3d')),

            meta(
              name_ : SFString('generator'),
              content_ : SFString('X3D-Edit 3.3, https://savage.nps.edu/X3D-Edit')),

            meta(
              name_ : SFString('license'),
              content_ : SFString('../license.html'))]),
      Scene_ : 
        Scene(
          children_ : [
            Transform(
              translation_ : SFVec3f([SFDouble(0), SFDouble(2), SFDouble(0)]),
              child_ : 
                Shape(
                  geometry_ : 
                    Text(
                      string_ : MFString([SFString("Compare special character escaping")]),
                      fontStyle_ : 
                        FontStyle(
                          DEF_ : SFString('testFontStyle'),
                          justify_ : MFString([SFString("MIDDLE"), SFString("MIDDLE")]),
                          size_ : 0.8)),
                  appearance_ : 
                    Appearance(
                      DEF_ : SFString('LightBlueAppearance'),
                      material_ : 
                        Material(
                          diffuseColor_ : SFColor([SFDouble(0.1), SFDouble(0.7), SFDouble(0.7)]))))),

            Transform(
              translation_ : SFVec3f([SFDouble(-3), SFDouble(0), SFDouble(0)]),
              child_ : 
                Shape(
                  geometry_ : 
                    Text(
                      string_ : MFString([SFString("I don't think so"), SFString(""), SFString("he said \"Hi\"")]),
                      fontStyle_ : 
                        FontStyle(
                          USE_ : SFString('testFontStyle'))),
                  appearance_ : 
                    Appearance(
                      USE_ : SFString('LightBlueAppearance')))),

            Transform(
              translation_ : SFVec3f([SFDouble(3), SFDouble(0), SFDouble(0)]),
              child_ : 
                Shape(
                  geometry_ : 
                    Text(
                      string_ : MFString([SFString("I don't think so"), SFString(""), SFString("he said \"Hi\"")]),
                      fontStyle_ : 
                        FontStyle(
                          USE_ : SFString('testFontStyle'))),
                  appearance_ : 
                    Appearance(
                      USE_ : SFString('LightBlueAppearance'))))]));
void main() { exit(0); }

File: src/main/go/net/coderextreme/data/TextExamples.go

package main

import (
	"encoding/xml"
	"fmt"
	"io"
	"log"
	"net/http"
	"os"

	"gox3d/x3d" // Assumes your module is named x3d-go-tester

	// Imports for the CGo-based libxml2 wrapper
	"github.com/lestrrat-go/libxml2"
	"github.com/lestrrat-go/libxml2/xsd"
)

// ... (Helper functions remain the same) ...
func stringPtr(s string) *string { return &s }
func floatPtr(f float32) *float32 { return &f }
func doublePtr(d float64) *float64 { return &d }
func boolPtr(b bool) *bool       { return &b }
func int32Ptr(i int32) *int32    { return &i }

func downloadSchemaIfNotExists(url, filepath string) error {
	// ... (function is correct, no changes) ...
	if _, err := os.Stat(filepath); err == nil {
		fmt.Printf("Schema file '%s' already exists. Skipping download.\n", filepath)
		return nil
	}
	fmt.Printf("Downloading X3D schema from %s to %s...\n", url, filepath)
	resp, err := http.Get(url)
	if err != nil { return fmt.Errorf("failed to download schema: %w", err) }
	defer resp.Body.Close()
	if resp.StatusCode != http.StatusOK { return fmt.Errorf("bad status: %s", resp.Status) }
	out, err := os.Create(filepath)
	if err != nil { return fmt.Errorf("failed to create schema file: %w", err) }
	defer out.Close()
	_, err = io.Copy(out, resp.Body)
	if err != nil { return fmt.Errorf("failed to save schema: %w", err) }
	fmt.Println("Schema downloaded successfully.")
	return nil
}

// Updated validation function to print detailed errors
func validateXMLWithSchema(xmlData []byte, schemaPath string) error {
	schema, err := xsd.ParseFromFile(schemaPath)
	if err != nil { return fmt.Errorf("failed to parse schema file %s: %w", schemaPath, err) }
	defer schema.Free()
	doc, err := libxml2.Parse(xmlData)
	if err != nil { return fmt.Errorf("failed to parse XML data: %w", err) }
	defer doc.Free()

	validationErr := schema.Validate(doc)
	if validationErr != nil {
		errorLog := ""
		if schemaErrors, ok := validationErr.(xsd.SchemaValidationError); ok {
			for i, e := range schemaErrors.Errors() {
				errorLog += fmt.Sprintf("Error %d: %s\n", i+1, e.Error())
			}
			return fmt.Errorf("schema validation failed with %d errors:\n%s", len(schemaErrors.Errors()), errorLog)
		}
		return fmt.Errorf("schema validation failed: %w", validationErr)
	}
	return nil
}

func main() {
	fmt.Println("--- Building and Testing an X3D Scene in Go ---")

	const schemaURL = "https://www.web3d.org/specifications/x3d-4.0.xsd"
	const schemaFilename = "x3d-4.0.xsd"
	if err := downloadSchemaIfNotExists(schemaURL, schemaFilename); err != nil {
		log.Fatalf("Could not prepare schema file: %v", err)
	}
    sceneRoot := &x3d.X3D{
        Profile: stringPtr("Immersive"),
        Version: stringPtr("4.0"),
        Head: &x3d.Head{
            Metas: []*x3d.Meta{
                &x3d.Meta{
                    Name: stringPtr("title"),
                    Content: stringPtr("TextExamples.x3d"),
            },
            &x3d.Meta{
                Name: stringPtr("description"),
                Content: stringPtr("Show different escape-character text examples for embedded quotation marks."),
            },
            &x3d.Meta{
                Name: stringPtr("creator"),
                Content: stringPtr("Don Brutzman"),
            },
            &x3d.Meta{
                Name: stringPtr("created"),
                Content: stringPtr("7 April 2001"),
            },
            &x3d.Meta{
                Name: stringPtr("modified"),
                Content: stringPtr("Tue, 09 Sep 2025 19:39:22 GMT"),
            },
            &x3d.Meta{
                Name: stringPtr("warning"),
                Content: stringPtr("Note that X3D Canonicalization (C14N) will scrub alternate XML character representations, be careful to check original encoding into version control."),
            },
            &x3d.Meta{
                Name: stringPtr("warning"),
                Content: stringPtr("Usually this source document needs to be inspected and edited using a plain-text editor in order to see the differences in these XML-equivalent text representations."),
            },
            &x3d.Meta{
                Name: stringPtr("identifier"),
                Content: stringPtr("https://www.web3d.org/x3d/content/examples/Basic/development/TextExamples.x3d"),
            },
            },
        },
        Scene: &x3d.Scene{
            Children: []x3d.X3DChildNode{
                &x3d.Transform{
                        Translation: &x3d.SFVec3f{0.0, 2.0, 0.0},
                    Children: []x3d.X3DNode{
                        &x3d.Shape{
                            Appearance: &x3d.Appearance{
                                CoreX3DNode: x3d.CoreX3DNode{
                                    DEF: stringPtr("LightBlueAppearance"),
                                },
                                Material: &x3d.Material{
                                    DiffuseColor: &x3d.SFColor{0.1, 0.7, 0.7},
                                },
                            },
                            Geometry: &x3d.Text{
                                String: x3d.MFString{"Compare special character escaping"},
                                FontStyle: &x3d.FontStyle{
                                    CoreX3DNode: x3d.CoreX3DNode{
                                        DEF: stringPtr("testFontStyle"),
                                    },
                                    Size: floatPtr(0.8),
                                    Justify: x3d.MFString{"MIDDLE", "MIDDLE"},
                                },
                            },
                        },
                    },
                },
                &x3d.Transform{
                        Translation: &x3d.SFVec3f{-3.0, 0.0, 0.0},
                    Children: []x3d.X3DNode{
                        &x3d.Shape{
                            Appearance: &x3d.Appearance{
                                CoreX3DNode: x3d.CoreX3DNode{
                                    USE: stringPtr("LightBlueAppearance"),
                                },
                            },
                            Geometry: &x3d.Text{
                                String: x3d.MFString{"I don't think so", "", "he said \"Hi\""},
                                FontStyle: &x3d.FontStyle{
                                    CoreX3DNode: x3d.CoreX3DNode{
                                        USE: stringPtr("testFontStyle"),
                                    },
                                },
                            },
                        },
                    },
                },
                &x3d.Transform{
                        Translation: &x3d.SFVec3f{3.0, 0.0, 0.0},
                    Children: []x3d.X3DNode{
                        &x3d.Shape{
                            Appearance: &x3d.Appearance{
                                CoreX3DNode: x3d.CoreX3DNode{
                                    USE: stringPtr("LightBlueAppearance"),
                                },
                            },
                            Geometry: &x3d.Text{
                                String: x3d.MFString{"I don't think so", "", "he said \"Hi\""},
                                FontStyle: &x3d.FontStyle{
                                    CoreX3DNode: x3d.CoreX3DNode{
                                        USE: stringPtr("testFontStyle"),
                                    },
                                },
                            },
                        },
                    },
                },
            },
        },
    }


	fmt.Println("\n--- Validating the generated scene (internal logic) ---")
	err := sceneRoot.Scene.Validate()
	if err != nil {
		log.Fatalf("Scene validation failed: %v", err)
	}
	fmt.Println("✅ Internal scene validation successful!")
	fmt.Println("\n--- Serializing scene to XML ---")
	output, err := xml.MarshalIndent(sceneRoot, "", "  ")

	if err != nil {
		log.Fatalf("XML Marshaling failed: %v", err)
	}
	/*
	fmt.Println("\n--- Validating XML against X3D 4.0 Schema (using libxml2) ---")
	err = validateXMLWithSchema(output, schemaFilename)
	if err != nil {
		fmt.Printf("--- Invalid Generated XML ---\n%s\n---------------------------\n", string(output))
		log.Fatalf("Schema validation failed for generated XML: %v", err)
	}
	fmt.Println("✅ XML is valid against the X3D 4.0 schema!")
	*/
	filename := "../data/TextExamples.new.go.x3d"
	file, err := os.OpenFile(filename, os.O_CREATE|os.O_WRONLY, 0644)
	if err != nil {
		fmt.Printf("Error opening file: %v\n", err)
		return
	}
	defer file.Close() // Ensure the file is closed when the function exits

	// Write the string content to the file
	header := "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n<!DOCTYPE X3D PUBLIC \"ISO//Web3D//DTD X3D 4.0//EN\" \"https://www.web3d.org/specifications/x3d-4.0.dtd\">\n"
	_, err = file.WriteString(header)
	if err != nil {
		fmt.Printf("Error writing header to file: %v\n", err)
		return
	}

	_, err = file.Write(output)
	if err != nil {
		fmt.Printf("Error writing bytes to file: %v\n", err)
		return
	}

	fmt.Printf("\n✅ Scene successfully written to %s\n", filename)
	fmt.Printf("   You can view the generated XML below:\n\n%s\n", string(output))
}